File: solutions/kotlin/resistor-color-trio/2/src/main/kotlin/ResistorColorTrio.kt
import kotlin.math.pow

object ResistorColorTrio {
    private val UNITS = arrayOf("ohms", "kiloohms", "megaohms", "gigaohms")

    fun text(vararg input: Color): String {
        require(input.size >= 3) { "Must provide at least 3 colors" }

        val (first, second, third) = input
        val baseValue = first.ordinal * 10 + second.ordinal
        val multiplier = 10.0.pow(third.ordinal).toInt()

        val resistance = (baseValue * multiplier)

        var numeral = resistance.toDouble()
        var unitIndex = 0
        
        while (numeral >= 1000 && unitIndex < UNITS.lastIndex) {
            numeral /= 1000
            ++unitIndex
        }

        val formatted = if (numeral % 1 == 0.0)
                            numeral.toInt().toString()
                        else "%.1f".format(numeral)
                        
        return "$formatted ${UNITS[unitIndex]}"
    }
}

File: solutions/c/resistor-color-trio/2/resistor_color_trio.c
#include "resistor_color_trio.h"
#include <math.h>

resistor_value_t color_code(const resistor_band_t colors[static 3]) {
    unit_t unit = colors[2] / 3;
    uint16_t value = (10 * colors[0] + colors[1]) * pow(10, colors[2] % 3);
    while (value >= 1000) {
        value /= 1000;
        ++unit;
    }
    return (resistor_value_t) {
        .unit = unit,
        .value = value,
    };
}